File: core/tests/unit/conformance.cpp
#include <cassert>
#include <filesystem>
#include <fstream>
#include <iterator>

#include "streamfind/project.hpp"
#include "../tmp_projects.hpp"

namespace streamfind::conformance {

Json fixture() {
    std::ifstream input(STREAMFIND_CONFORMANCE_FIXTURE);
    return Json::parse(std::string(std::istreambuf_iterator<char>(input), {}));
}

void install_domain_schema(const Project &project) {
    static constexpr const char *tables[] = {
        "MASS_SPEC_ANALYSES", "MASS_SPEC_SPECTRA_HEADERS",
        "MASS_SPEC_CHROMATOGRAMS_HEADERS", "MASS_SPEC_CHROMATOGRAMS",
        "MASS_SPEC_NTA_FEATURES", "MASS_SPEC_NTA_SUSPECTS",
        "MASS_SPEC_NTA_INTERNAL_STANDARDS", "MASS_SPEC_NTA_TRANSFORMATION_PRODUCTS"};
    for (const auto *table : tables)
        project.execute_sql(std::string("CREATE TABLE ") + table + " (value VARCHAR)");
}

std::filesystem::path temporary_database(const char *name) {
    auto path = streamfind::test::tmp_projects_dir() / name;
    std::error_code error;
    std::filesystem::remove(path, error);
    return path;
}

void assert_project(const Project &project, const Json &expected) {
    assert(project.get_domain() == expected.at("domain").get<std::string>());
    assert(project.get_metadata() == expected.at("metadata"));
    assert(project.get_workflow().to_json() == expected.at("workflow"));
    assert(project.get_cache_size() == 1);
    assert(project.get_cache().front().hash == expected.at("cache").at("hash").get<std::string>());
}

void run() {
    const Json expected = fixture();
    const auto path = temporary_database("streamfind-cpp-conformance.duckdb");
    auto project = Project::create({path, expected.at("domain"), expected.at("metadata")});
    project.set_metadata(expected.at("metadata"));
    project.set_workflow(Workflow::from_json(expected.at("workflow")));
    project.set_cache(expected.at("cache").at("name"), expected.at("cache").at("description"),
                      expected.at("cache").at("hash"), expected.at("cache").at("value"));
    install_domain_schema(project);
    assert_project(project, expected);
    project.validate();
    project.close();

    auto reopened = Project::open({path, expected.at("domain"), {}});
    assert_project(reopened, expected);
    reopened.validate();
    reopened.close();
    std::error_code error;
    std::filesystem::remove(path, error);
}

} // namespace streamfind::conformance

int main() {
    streamfind::conformance::run();
    return 0;
}

File: rust/crates/core/tests/conformance.rs
use serde_json::Value;
use std::fs;
use std::path::PathBuf;
use streamfind_rust_core::{Project, ProjectOptions};

fn fixture() -> Value {
    serde_json::from_str(include_str!(
        "../../../../tests/fixtures/project/project_conformance.json"
    ))
    .unwrap()
}

fn domain_schema_fixture() -> Value {
    serde_json::from_str(include_str!(
        "../../../../tests/fixtures/project/domain_schema_manifest.json"
    ))
    .unwrap()
}

fn database(name: &str) -> PathBuf {
    let path = streamfind_rust_test_support::tmp_projects_dir().join(name);
    let _ = fs::remove_file(&path);
    path
}

fn install_domain_schema(project: &mut Project) {
    for table in domain_schema_fixture()["domains"]["mass_spec"]["required_tables"]
        .as_array()
        .unwrap()
    {
        project
            .execute_sql(&format!(
                "CREATE TABLE {} (marker VARCHAR)",
                table.as_str().unwrap()
            ))
            .unwrap();
    }
}

fn options(path: PathBuf, fixture: &Value, read_only: bool) -> ProjectOptions {
    ProjectOptions {
        database_path: path,
        domain: fixture["domain"].as_str().unwrap().into(),
        create_if_missing: false,
        read_only,
    }
}

#[test]
fn shared_fixture_round_trips_project_contract() {
    let fixture = fixture();
    let path = database("streamfind-rust-conformance.duckdb");
    let mut project = Project::create(options(path.clone(), &fixture, false)).unwrap();
    project.set_metadata(fixture["metadata"].clone()).unwrap();
    project
        .set_workflow(
            streamfind_rust_core::Workflow::from_json(&fixture["workflow"]).unwrap(),
            &streamfind_rust_core::MethodRegistry::default(),
        )
        .unwrap();
    project
        .set_cache(
            fixture["cache"]["name"].as_str().unwrap(),
            fixture["cache"]["description"].as_str().unwrap(),
            fixture["cache"]["hash"].as_str().unwrap(),
            &fixture["cache"]["value"],
        )
        .unwrap();
    install_domain_schema(&mut project);
    assert_eq!(project.get_domain(), fixture["domain"]);
    assert_eq!(project.get_metadata(), fixture["metadata"]);
    assert_eq!(
        project.get_workflow().unwrap().to_json(),
        fixture["workflow"]
    );
    assert_eq!(project.get_cache_size().unwrap(), 1);
    project.validate().unwrap();
    drop(project);

    let reopened = Project::open(options(path.clone(), &fixture, true)).unwrap();
    assert_eq!(
        reopened.get_cache().unwrap()[0].hash,
        fixture["cache"]["hash"]
    );
    reopened.validate().unwrap();
    drop(reopened);
    fs::remove_file(path).unwrap();
}
